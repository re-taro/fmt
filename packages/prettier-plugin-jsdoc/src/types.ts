import type { ParserOptions } from "prettier";

export interface JsdocOptions {
  jsdocSpaces: number;
  jsdocPrintWidth?: number;
  jsdocDescriptionWithDot: boolean;
  jsdocDescriptionTag: boolean;
  jsdocVerticalAlignment: boolean;
  jsdocKeepUnParseAbleExampleIndent: boolean;
  /** @default "singleLine" */
  jsdocCommentLineStrategy: "singleLine" | "multiline" | "keep";
  jsdocSeparateReturnsFromParam: boolean;
  jsdocSeparateTagGroups: boolean;
  jsdocAddDefaultToDescription: boolean;
  jsdocCapitalizeDescription: boolean;
  jsdocPreferCodeFences: boolean;
  tsdoc: boolean;
  jsdocLineWrappingStyle: "greedy";
}

export interface AllOptions extends ParserOptions, JsdocOptions {}

interface LocationDetails {
  line: number;
  column: number;
}
interface Location {
  start: LocationDetails;
  end: LocationDetails;
}

export interface PrettierComment {
  type: "CommentBlock" | "Block";
  value: string;
  start: number;
  end: number;
  loc: Location;
}

export interface Token {
  type:
    | "CommentBlock"
    | "Block"
    | {
        label: string;
        keyword?: string;
        beforeExpr: boolean;
        startsExpr: boolean;
        rightAssociative: boolean;
        isLoop: boolean;
        isAssign: boolean;
        prefix: boolean;
        postfix: boolean;
        binop: null;
      };
  value: string;
  start: number;
  end: number;
  loc: Location;
}

export interface AST {
  start: number;
  end: number;
  loc: Location;
  errors: [];
  program: {
    type: "Program";
    start: number;
    end: number;
    loc: [];
    sourceType: "module";
    interpreter: null;
    body: [];
    directives: [];
  };
  comments: PrettierComment[];
  tokens: Token[];
}
