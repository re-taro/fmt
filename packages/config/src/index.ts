export default {
  extends: [
    "plugin:@re-taro/json",
    "plugin:@re-taro/markdown",
    "plugin:@re-taro/react",
    "plugin:@re-taro/svelte",
    "plugin:@re-taro/typescript",
    "plugin:@re-taro/vue",
    "plugin:@re-taro/yaml",
    "prettier",
  ],
};
