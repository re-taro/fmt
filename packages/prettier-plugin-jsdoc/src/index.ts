import type { Parser, ParserOptions, SupportOption } from "prettier";
import parserBabel from "prettier/plugins/babel";
import parserFlow from "prettier/plugins/flow";
import parserTypescript from "prettier/plugins/typescript";

import { getParser } from "./parser";
import type { JsdocOptions } from "./types";
import { findPluginByParser } from "./utils";

export const options = {
  jsdocSpaces: {
    name: "jsdocSpaces",
    type: "int",
    category: "jsdoc",
    default: 1,
    description: "How many spaces will be used to separate tag elements.",
  },
  jsdocDescriptionWithDot: {
    name: "jsdocDescriptionWithDot",
    type: "boolean",
    category: "jsdoc",
    default: false,
    description: "Should dot be inserted at the end of description",
  },
  jsdocDescriptionTag: {
    name: "jsdocDescriptionTag",
    type: "boolean",
    category: "jsdoc",
    default: false,
    description: "Should description tag be used",
  },
  jsdocVerticalAlignment: {
    name: "jsdocVerticalAlignment",
    type: "boolean",
    category: "jsdoc",
    default: false,
    description: "Should tags, types, names and description be aligned",
  },
  jsdocKeepUnParseAbleExampleIndent: {
    name: "jsdocKeepUnParseAbleExampleIndent",
    type: "boolean",
    category: "jsdoc",
    default: false,
    description:
      "Should unParseAble example (pseudo code or no js code) keep its indentation",
  },
  jsdocCommentLineStrategy: {
    name: "jsdocCommentLineStrategy",
    type: "choice",
    choices: [
      {
        since: "1.1.0",
        value: "singleLine",
        description: "Should compact single line comment, if possible",
      },
      {
        since: "1.1.0",
        value: "multiline",
        description: "Should compact multi line comment",
      },
      {
        since: "1.1.0",
        value: "keep",
        description: "Should keep original line comment",
      },
    ],
    category: "jsdoc",
    default: "singleLine",
    description: "How comments line should be",
  },
  jsdocSeparateReturnsFromParam: {
    name: "jsdocSeparateReturnsFromParam",
    type: "boolean",
    category: "jsdoc",
    default: false,
    description: "Add an space between last @param and @returns",
  },
  jsdocSeparateTagGroups: {
    name: "jsdocSeparateTagGroups",
    type: "boolean",
    category: "jsdoc",
    default: false,
    description: "Add an space between tag groups",
  },
  jsdocCapitalizeDescription: {
    name: "jsdocCapitalizeDescription",
    type: "boolean",
    category: "jsdoc",
    default: true,
    description: "Should capitalize first letter of description",
  },
  tsdoc: {
    name: "tsdoc",
    type: "boolean",
    category: "jsdoc",
    default: false,
    description: "Should format as tsdoc",
  },
  jsdocPrintWidth: {
    name: "jsdocPrintWidth",
    type: "int",
    category: "jsdoc",
    default: undefined as any,
    description:
      "If You don't set value to jsdocPrintWidth, the printWidth will be use as jsdocPrintWidth.",
  },
  jsdocAddDefaultToDescription: {
    name: "jsdocAddDefaultToDescription",
    type: "boolean",
    category: "jsdoc",
    default: true,
    description: "Add Default value of a param to end description",
  },
  jsdocPreferCodeFences: {
    name: "jsdocPreferCodeFences",
    type: "boolean",
    category: "jsdoc",
    default: false,
    description: `Prefer to render code blocks using "fences" (triple backticks). If not set, blocks without a language tag will be rendered with a four space indentation.`,
  },
  jsdocLineWrappingStyle: {
    name: "jsdocLineWrappingStyle",
    type: "choice",
    choices: [
      {
        since: "0.3.39",
        value: "greedy",
        description: "Lines wrap as soon as they reach the print width",
      },
    ],
    category: "jsdoc",
    default: "greedy",
    description:
      "Strategy for wrapping lines for the given print width. More options may be added in the future.",
  },
} satisfies Record<keyof JsdocOptions, SupportOption>;

export const defaultOptions: JsdocOptions = {
  jsdocSpaces: options.jsdocSpaces.default,
  jsdocPrintWidth: options.jsdocPrintWidth.default as unknown as undefined,
  jsdocDescriptionWithDot: options.jsdocDescriptionWithDot.default as boolean,
  jsdocDescriptionTag: options.jsdocDescriptionTag.default as boolean,
  jsdocVerticalAlignment: options.jsdocVerticalAlignment.default as boolean,
  jsdocKeepUnParseAbleExampleIndent: options.jsdocKeepUnParseAbleExampleIndent
    .default as boolean,
  jsdocCommentLineStrategy: options.jsdocCommentLineStrategy
    .default as "singleLine",
  jsdocSeparateReturnsFromParam: options.jsdocSeparateReturnsFromParam
    .default as boolean,
  jsdocSeparateTagGroups: options.jsdocSeparateTagGroups.default as boolean,
  jsdocCapitalizeDescription: options.jsdocCapitalizeDescription
    .default as boolean,
  jsdocAddDefaultToDescription: options.jsdocAddDefaultToDescription
    .default as boolean,
  jsdocPreferCodeFences: options.jsdocPreferCodeFences.default as boolean,
  tsdoc: options.tsdoc.default as boolean,
  jsdocLineWrappingStyle: options.jsdocLineWrappingStyle.default as "greedy",
};

export const parsers = {
  // JS - Babel
  get babel() {
    const parser = parserBabel.parsers.babel;

    return mergeParsers(parser, "babel");
  },
  get "babel-flow"() {
    const parser = parserBabel.parsers["babel-flow"];

    return mergeParsers(parser, "babel-flow");
  },
  get "babel-ts"() {
    const parser = parserBabel.parsers["babel-ts"];

    return mergeParsers(parser, "babel-ts");
  },
  // JS - Flow
  get flow() {
    const parser = parserFlow.parsers.flow;

    return mergeParsers(parser, "flow");
  },
  // JS - TypeScript
  get typescript(): Parser {
    const parser = parserTypescript.parsers.typescript;

    return mergeParsers(parser, "typescript");
    // require("./parser-typescript").parsers.typescript;
  },
  get "jsdoc-parser"() {
    // Backward compatible, don't use this in new version since 1.0.0
    const parser = parserBabel.parsers["babel-ts"];

    return mergeParsers(parser, "babel-ts");
  },
};

function mergeParsers(originalParser: Parser, parserName: string) {
  const jsDocParse = getParser(originalParser.parse, parserName) as any;

  const parser = {
    ...originalParser,
    preprocess: jsDocPreprocess,
    parse: jsDocParse,
  };

  function jsDocPreprocess(text: string, options: ParserOptions) {
    const tsPluginParser = findPluginByParser(parserName, options);

    if (!tsPluginParser) {
      return originalParser.preprocess
        ? originalParser.preprocess(text, options)
        : text;
    }

    const preprocess = tsPluginParser.preprocess ?? originalParser.preprocess;

    Object.assign(parser, {
      ...parser,
      ...tsPluginParser,
      preprocess: jsDocPreprocess,
      parse: jsDocParse,
    });

    return preprocess ? preprocess(text, options) : text;
  }

  return parser;
}
export type Options = Partial<JsdocOptions>;
