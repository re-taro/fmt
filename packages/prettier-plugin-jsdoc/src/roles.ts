import {
  ABSTRACT,
  ASYNC,
  AUGMENTS,
  AUTHOR,
  BORROWS,
  CALLBACK,
  CATEGORY,
  CLASS,
  CONSTANT,
  DEFAULT,
  DEFAULT_VALUE,
  DEPRECATED,
  DESCRIPTION,
  EXAMPLE,
  EXTENDS,
  EXTERNAL,
  FILE,
  FIRES,
  FLOW,
  FUNCTION,
  IGNORE,
  LICENSE,
  MEMBER,
  MEMBEROF,
  MODULE,
  NAMESPACE,
  OVERLOAD,
  OVERRIDE,
  PARAM,
  PRIVATE,
  PRIVATE_REMARKS,
  PROPERTY,
  PROVIDES_MODULE,
  REMARKS,
  RETURNS,
  SATISFIES,
  SEE,
  SINCE,
  TEMPLATE,
  THROWS,
  TODO,
  TYPE,
  TYPE_PARAM,
  TYPEDEF,
  VERSION,
  YIELDS,
} from "./tags";

const TAGS_SYNONYMS = {
  arg: PARAM,
  argument: PARAM,
  const: CONSTANT,
  constructor: CLASS,
  desc: DESCRIPTION,
  emits: FIRES,
  examples: EXAMPLE,
  exception: THROWS,
  fileoverview: FILE,
  func: FUNCTION,
  host: EXTERNAL,
  method: FUNCTION,
  overview: FILE,
  params: PARAM,
  prop: PROPERTY,
  return: RETURNS,
  var: MEMBER,
  virtual: ABSTRACT,
  yield: YIELDS,
  hidden: IGNORE,
};

const TAGS_DEFAULT = [DEFAULT, DEFAULT_VALUE];

const TAGS_NAMELESS = [
  BORROWS,
  CATEGORY,
  DEPRECATED,
  DESCRIPTION,
  EXAMPLE,
  EXTENDS,
  LICENSE,
  MODULE,
  NAMESPACE,
  OVERLOAD,
  OVERRIDE,
  PRIVATE_REMARKS,
  REMARKS,
  RETURNS,
  SINCE,
  THROWS,
  TODO,
  YIELDS,
  FILE,
  ...TAGS_DEFAULT,
];

const TAGS_TYPELESS = [
  BORROWS,
  BORROWS,
  DEPRECATED,
  DESCRIPTION,
  EXAMPLE,
  IGNORE,
  LICENSE,
  MODULE,
  NAMESPACE,
  OVERLOAD,
  OVERRIDE,
  PRIVATE_REMARKS,
  REMARKS,
  SINCE,
  TODO,
  FILE,
];

const TAGS_PEV_FORMATE_DESCRIPTION = [
  BORROWS,
  ...TAGS_DEFAULT,
  MEMBEROF,
  MODULE,
  SEE,
];

const TAGS_DESCRIPTION_NEEDED = [
  BORROWS,
  CATEGORY,
  DESCRIPTION,
  EXAMPLE,
  PRIVATE_REMARKS,
  REMARKS,
  SINCE,
  TODO,
];

const TAGS_TYPE_NEEDED = [
  EXTENDS,
  PARAM,
  PROPERTY,
  RETURNS,
  THROWS,
  TYPE,
  SATISFIES,
  TYPEDEF,
  YIELDS,
];

const TAGS_VERTICALLY_ALIGN_ABLE = [
  EXTENDS,
  PARAM,
  PROPERTY,
  RETURNS,
  THROWS,
  TYPE,
  SATISFIES,
  TYPEDEF,
  YIELDS,
];

const TAGS_GROUP_HEAD = [CALLBACK, TYPEDEF];
const TAGS_GROUP_CONDITION = [
  ...TAGS_GROUP_HEAD,
  TYPE,
  PROPERTY,
  PARAM,
  RETURNS,
  YIELDS,
  THROWS,
];

const TAGS_ORDER = [
  REMARKS,
  PRIVATE_REMARKS,
  PROVIDES_MODULE,
  MODULE,
  LICENSE,
  FLOW,
  ASYNC,
  PRIVATE,
  IGNORE,
  MEMBEROF,
  VERSION,
  FILE,
  AUTHOR,
  DEPRECATED,
  SINCE,
  CATEGORY,
  DESCRIPTION,
  EXAMPLE,
  ABSTRACT,
  AUGMENTS,
  CONSTANT,
  ...TAGS_DEFAULT,
  EXTERNAL,
  OVERLOAD,
  FIRES,
  TEMPLATE,
  TYPE_PARAM,
  FUNCTION,
  NAMESPACE,
  BORROWS,
  CLASS,
  EXTENDS,
  MEMBER,
  TYPEDEF,
  TYPE,
  SATISFIES,
  PROPERTY,
  CALLBACK,
  PARAM,
  YIELDS,
  RETURNS,
  THROWS,
  "other",
  SEE,
  TODO,
];

export {
  TAGS_DEFAULT,
  TAGS_DESCRIPTION_NEEDED,
  TAGS_GROUP_CONDITION,
  TAGS_GROUP_HEAD,
  TAGS_NAMELESS,
  TAGS_ORDER,
  TAGS_PEV_FORMATE_DESCRIPTION,
  TAGS_SYNONYMS,
  TAGS_TYPE_NEEDED,
  TAGS_TYPELESS,
  TAGS_VERTICALLY_ALIGN_ABLE,
};
