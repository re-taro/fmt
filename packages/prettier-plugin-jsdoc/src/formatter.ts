import type { Image, Link, List, Root, RootContent, Text } from "mdast";
import { fromMarkdown } from "mdast-util-from-markdown";
import type { BuiltInParserName } from "prettier";
import { format } from "prettier";

import { TAGS_PEV_FORMATE_DESCRIPTION } from "./roles";
import { DESCRIPTION, EXAMPLE, TODO } from "./tags";
import type { AllOptions } from "./types";
import { capitalizer, formatCode } from "./utils";

const TABLE = "2@^5!~#sdE!_TABLE";

interface DescriptionEndLineParams {
  tag: string;
  isEndTag: boolean;
}

function parserSynonyms(lang: string): BuiltInParserName[] {
  switch (lang) {
    case "js":
    case "javascript":
    case "jsx": {
      return ["babel", "babel-flow", "vue"];
    }
    case "ts":
    case "typescript":
    case "tsx": {
      return ["typescript", "babel-ts", "angular"];
    }
    case "json":
    case "css": {
      return ["css"];
    }
    case "less": {
      return ["less"];
    }
    case "scss": {
      return ["scss"];
    }
    case "html": {
      return ["html"];
    }
    case "yaml": {
      return ["yaml"];
    }
    default: {
      return ["babel"];
    }
  }
}

function descriptionEndLine({
  tag,
  isEndTag,
}: DescriptionEndLineParams): string {
  if ([DESCRIPTION, EXAMPLE, TODO].includes(tag) && !isEndTag) {
    return "\n";
  }

  return "";
}

interface FormatOptions {
  tagStringLength?: number;
  beginningSpace: string;
}

/**
 * Trim, make single line with capitalized text. Insert dot if flag for it is
 * set to true and last character is a word character
 *
 * @private
 */
async function formatDescription(
  tag: string,
  text: string,
  options: AllOptions,
  formatOptions: FormatOptions,
): Promise<string> {
  if (!text) return text;

  const { printWidth } = options;
  const { tagStringLength = 0, beginningSpace } = formatOptions;

  text = text.replace(/^(\d+)[-][\s|]+/g, "$1. "); // Start
  text = text.replace(/\n+(\s*\d+)[-][\s]+/g, "\n$1. ");

  const fencedCodeBlocks = text.matchAll(/```\S*?\n[\s\S]+?```/gm);
  const indentedCodeBlocks = text.matchAll(
    /^\r?\n^(?:(?:(?:[ ]{4}|\t).*(?:\r?\n|$))+)/gm,
  );
  const allCodeBlocks = [...fencedCodeBlocks, ...indentedCodeBlocks];
  const tables: string[] = [];
  text = text.replace(
    /((\n|^)\|[\s\S]*?)((\n[^|])|$)/g,
    (code, _1, _2, _3, _, offs: number) => {
      for (const block of allCodeBlocks) {
        if (
          block.index !== undefined &&
          block.index <= offs + 1 &&
          offs + code.length + 1 <= block.index + block[0].length
        ) {
          return code;
        }
      }

      code = _3 ? code.slice(0, -1) : code;

      tables.push(code);

      return `\n\n${TABLE}\n\n${_3 ? _3.slice(1) : ""}`;
    },
  );

  if (
    options.jsdocCapitalizeDescription &&
    !TAGS_PEV_FORMATE_DESCRIPTION.includes(tag)
  ) {
    text = capitalizer(text);
  }

  text = `${tagStringLength ? `${"!".repeat(tagStringLength - 1)}?` : ""}${
    text.startsWith("```") ? "\n" : ""
  }${text}`;

  let tableIndex = 0;

  text = text.replace(
    new RegExp("\\n" + `[\u0020]{${beginningSpace.length}}`, "g"),
    "\n",
  );

  const rootAst = fromMarkdown(text);

  async function stringifyASTWithoutChildren(
    mdAst: RootContent | Root,
    intention: string,
    parent: RootContent | Root | null,
  ) {
    if (mdAst.type === "inlineCode") {
      return `\`${mdAst.value}\``;
    }

    if (mdAst.type === "code") {
      let result = mdAst.value || "";
      let _intention = intention;

      if (result) {
        if (mdAst.lang) {
          const supportParsers = parserSynonyms(mdAst.lang.toLowerCase());
          const parser = supportParsers?.includes(options.parser as any)
            ? options.parser
            : supportParsers?.[0] || mdAst.lang;

          result = await formatCode(result, intention, {
            ...options,
            parser,
            jsdocKeepUnParseAbleExampleIndent: true,
          });
        } else if (options.jsdocPreferCodeFences || false) {
          result = await formatCode(result, _intention, {
            ...options,
            jsdocKeepUnParseAbleExampleIndent: true,
          });
        } else {
          _intention = intention + " ".repeat(4);

          result = await formatCode(result, _intention, {
            ...options,
            jsdocKeepUnParseAbleExampleIndent: true,
          });
        }
      }
      const addFence = options.jsdocPreferCodeFences || !!mdAst.lang;
      result = addFence ? result : result.trimEnd();

      return result
        ? addFence
          ? `\n\n${_intention}\`\`\`${mdAst.lang ?? ""}${result}\`\`\``
          : `\n${result}`
        : "";
    }

    if ((mdAst as Text).value === TABLE) {
      if (parent) {
        (parent as any).costumeType = TABLE;
      }

      if (tables.length > 0) {
        let result = tables?.[tableIndex] || "";
        tableIndex++;
        if (result) {
          result = (
            await format(result, {
              ...options,
              parser: "markdown",
            })
          ).trim();
        }

        return `${
          result
            ? `\n\n${intention}${result.split("\n").join(`\n${intention}`)}`
            : (mdAst as Text).value
        }`;
      }
    }

    if (mdAst.type === "break") {
      return "\\\n";
    }

    return (
      (mdAst as Text).value ??
      (mdAst as Link).title ??
      (mdAst as Image).alt ??
      ""
    );
  }

  async function stringyfy(
    mdAst: RootContent | Root,
    intention: string,
    parent: RootContent | Root | null,
  ): Promise<string> {
    if (!Array.isArray((mdAst as Root).children)) {
      return stringifyASTWithoutChildren(mdAst, intention, parent);
    }

    return (
      await Promise.all(
        (mdAst as Root).children.map(async (ast, index) => {
          switch (ast.type) {
            case "listItem": {
              let _listCount = `\n${intention}- `;
              if (typeof (mdAst as List).start === "number") {
                const count = index + ((mdAst as List).start ?? 1);
                _listCount = `\n${intention}${count}. `;
              }

              const _intention = intention + " ".repeat(_listCount.length - 1);

              const result = (await stringyfy(ast, _intention, mdAst)).trim();

              return `${_listCount}${result}`;
            }

            case "list": {
              let end = "";
              if (
                tag !== DESCRIPTION &&
                mdAst.type === "root" &&
                index === mdAst.children.length - 1
              ) {
                end = "\n";
              }

              return `\n${await stringyfy(ast, intention, mdAst)}${end}`;
            }

            case "paragraph": {
              const paragraph = await stringyfy(ast, intention, parent);
              if ((ast as any).costumeType === TABLE) {
                return paragraph;
              }

              return `\n\n${paragraph
                .split("\\\n")
                .map((_paragraph) => {
                  const links: string[] = [];
                  _paragraph = _paragraph.replace(
                    /{@(link|linkcode|linkplain)[\s]((?:[^{}])*)}/g,
                    (_, tag: string, link: string) => {
                      links.push(link);

                      return `{@${tag}${"_".repeat(link.length)}}`;
                    },
                  );

                  _paragraph = _paragraph.replace(/\s+/g, " ");

                  if (
                    options.jsdocCapitalizeDescription &&
                    !TAGS_PEV_FORMATE_DESCRIPTION.includes(tag)
                  )
                    _paragraph = capitalizer(_paragraph);
                  if (options.jsdocDescriptionWithDot)
                    _paragraph = _paragraph.replace(/([\w\p{L}])$/u, "$1.");

                  let result = breakDescriptionToLines(
                    _paragraph,
                    printWidth,
                    intention,
                  );

                  result = result.replace(
                    /{@(link|linkcode|linkplain)([_]+)}/g,
                    (original: string, tag: string, underline: string) => {
                      const link = links[0];

                      if (link.length === underline.length) {
                        links.shift();

                        return `{@${tag} ${link}}`;
                      }

                      return original;
                    },
                  );

                  return result;
                })
                .join("\\\n")}`;
            }

            case "strong": {
              return `**${await stringyfy(ast, intention, mdAst)}**`;
            }

            case "emphasis": {
              return `_${await stringyfy(ast, intention, mdAst)}_`;
            }

            case "heading": {
              return `\n\n${intention}${"#".repeat(
                ast.depth,
              )} ${await stringyfy(ast, intention, mdAst)}`;
            }

            case "link":
            case "image": {
              return `[${await stringyfy(ast, intention, mdAst)}](${ast.url})`;
            }

            case "linkReference": {
              return `[${await stringyfy(ast, intention, mdAst)}][${
                ast.label
              }]`;
            }
            case "definition": {
              return `\n\n[${ast.label}]: ${ast.url}`;
            }

            case "blockquote": {
              const paragraph = await stringyfy(ast, "", mdAst);

              return `${intention}> ${paragraph
                .trim()
                .replace(/(\n+)/g, `$1${intention}> `)}`;
            }
          }

          return stringyfy(ast, intention, mdAst);
        }),
      )
    ).join("");
  }

  let result = await stringyfy(rootAst, beginningSpace, null);

  result = result.replace(/^[\s\n]+/g, "");
  result = result.replace(/^(?:[!]+\?)/g, "");

  return result;
}

function breakDescriptionToLines(
  desContent: string,
  maxWidth: number,
  beginningSpace: string,
): string {
  let str = desContent.trim();

  if (!str) {
    return str;
  }

  let result = "";
  while (str.length > maxWidth) {
    let sliceIndex = str.lastIndexOf(
      " ",
      str.startsWith("\n") ? maxWidth + 1 : maxWidth,
    );
    if (sliceIndex <= beginningSpace.length)
      sliceIndex = str.indexOf(" ", beginningSpace.length + 1);

    if (sliceIndex === -1) sliceIndex = str.length;

    result += str.slice(0, Math.max(0, sliceIndex));
    str = str.slice(Math.max(0, sliceIndex + 1));
    if (str) {
      str = `${beginningSpace}${str}`;
      str = `\n${str}`;
    }
  }

  result += str;

  return `${beginningSpace}${result}`;
}

export type { FormatOptions };
export { descriptionEndLine, formatDescription };
