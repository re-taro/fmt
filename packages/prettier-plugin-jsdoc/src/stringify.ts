import type { Spec } from "comment-parser";

import { descriptionEndLine, formatDescription } from "./formatter";
import {
  TAGS_ORDER,
  TAGS_PEV_FORMATE_DESCRIPTION,
  TAGS_VERTICALLY_ALIGN_ABLE,
} from "./roles";
import {
  DESCRIPTION,
  EXAMPLE,
  PRIVATE_REMARKS,
  REMARKS,
  SPACE_TAG_DATA,
} from "./tags";
import type { AllOptions } from "./types";
import { formatCode, isDefaultTag } from "./utils";

async function stringify(
  { name, description, type, tag }: Spec,
  tagIndex: number,
  finalTagsArray: Spec[],
  options: AllOptions,
  maxTagTitleLength: number,
  maxTagTypeNameLength: number,
  maxTagNameLength: number,
): Promise<string> {
  let tagString = "\n";

  if (tag === SPACE_TAG_DATA.tag) {
    return tagString;
  }

  const {
    printWidth,
    jsdocSpaces,
    jsdocVerticalAlignment,
    jsdocDescriptionTag,
    tsdoc,
    useTabs,
    tabWidth,
    jsdocSeparateTagGroups,
  } = options;
  const gap = " ".repeat(jsdocSpaces);

  let tagTitleGapAdj = 0;
  let tagTypeGapAdj = 0;
  let tagNameGapAdj = 0;
  let descGapAdj = 0;

  if (jsdocVerticalAlignment && TAGS_VERTICALLY_ALIGN_ABLE.includes(tag)) {
    if (tag) tagTitleGapAdj += maxTagTitleLength - tag.length;
    else if (maxTagTitleLength) descGapAdj += maxTagTitleLength + gap.length;

    if (type) tagTypeGapAdj += maxTagTypeNameLength - type.length;
    else if (maxTagTypeNameLength)
      descGapAdj += maxTagTypeNameLength + gap.length;

    if (name) tagNameGapAdj += maxTagNameLength - name.length;
    else if (maxTagNameLength) descGapAdj = maxTagNameLength + gap.length;
  }

  const useTagTitle = tag !== DESCRIPTION || jsdocDescriptionTag;

  if (useTagTitle) {
    tagString += `@${tag}${" ".repeat(tagTitleGapAdj || 0)}`;
  }
  if (type) {
    function getUpdatedType() {
      if (!isDefaultTag(tag)) {
        return `{${type}}`;
      }

      if (type === "[]") return "[ ]";
      if (type === "{}") return "{ }";

      const isAnObject = (value: string): boolean =>
        /^{.*[A-z0-9_]+ ?:.*}$/.test(value);
      const fixObjectCommas = (objWithBrokenCommas: string): string =>
        objWithBrokenCommas.replace(/; ([A-z0-9_])/g, ", $1");

      if (isAnObject(type)) {
        return fixObjectCommas(type);
      }

      return type;
    }
    const updatedType = getUpdatedType();
    tagString += gap + updatedType + " ".repeat(tagTypeGapAdj);
  }
  if (name) tagString += `${gap}${name}${" ".repeat(tagNameGapAdj)}`;

  if (tag === EXAMPLE && !tsdoc) {
    const exampleCaption = description.match(
      /<caption>(?:[\s\S]*?)<\/caption>/i,
    );

    if (exampleCaption) {
      description = description.replace(exampleCaption[0], "");
      tagString = `${tagString} ${exampleCaption[0]}`;
    }

    const beginningSpace = useTabs ? "\t" : " ".repeat(tabWidth);
    const formattedExample = await formatCode(
      description,
      beginningSpace,
      options,
    );

    tagString += formattedExample
      .replace(
        new RegExp(
          `^\\n${beginningSpace
            .replace(/[\t]/g, "[\\t]")
            .replace(/[^S\r\n]/g, "[^S\\r\\n]")}\\n`,
        ),
        "",
      )
      .trimEnd();
  } else if (description) {
    let descriptionString = "";
    if (useTagTitle) tagString += gap + " ".repeat(descGapAdj);
    if (
      TAGS_PEV_FORMATE_DESCRIPTION.includes(tag) ||
      !TAGS_ORDER.includes(tag)
    ) {
      descriptionString = description;
    } else {
      const [, firstWord] = /^\s*(\S+)/.exec(description) ?? ["", ""];

      const beginningSpace =
        tag === DESCRIPTION ||
        ([EXAMPLE, REMARKS, PRIVATE_REMARKS].includes(tag) && tsdoc)
          ? ""
          : "  ";

      // eslint-disable-next-line unicorn/prefer-ternary
      if (
        (tag !== DESCRIPTION &&
          tagString.length + firstWord.length > printWidth) ||
        [REMARKS, PRIVATE_REMARKS].includes(tag)
      ) {
        descriptionString = `\n${beginningSpace}${await formatDescription(
          tag,
          description,
          options,
          {
            beginningSpace,
          },
        )}`;
      } else {
        descriptionString = await formatDescription(tag, description, options, {
          tagStringLength: tagString.length - 1,
          beginningSpace,
        });
      }
    }

    if (jsdocSeparateTagGroups) {
      descriptionString = descriptionString.trimEnd();
    }

    tagString += descriptionString.startsWith("\n")
      ? descriptionString.replace(/^\n[\s]+\n/g, "\n")
      : descriptionString.trimStart();
  }

  tagString += descriptionEndLine({
    tag,
    isEndTag: tagIndex === finalTagsArray.length - 1,
  });

  return tagString;
}

export { stringify };
