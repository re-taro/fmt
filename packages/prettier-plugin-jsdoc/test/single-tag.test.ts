import * as prettier from "prettier";
import { AllOptions } from "../src/types";
import { test, expect, describe } from "vitest";

function subject(code: string, options: Partial<AllOptions> = {}) {
  return prettier.format(code, {
    plugins: ["@re-taro/prettier-plugin-jsdoc"],
    jsdocSpaces: 1,
    parser: "babel",
    ...options,
  } as AllOptions);
}

test("single tag", async () => {
  const result = await subject(`
  /**
* @param {  string   }    param0 description
   */
function fun(param0){}

  export const SubDomain = {
    /**
     * @returns {import('axios').AxiosResponse<import('../types').SubDomain>}
     */
    async subDomain(subDomainAddress) {
    },
  };

`);

  expect(result).toMatchSnapshot();
});

describe("Comment Line Strategy", () => {
  test("keep single", async () => {
    const result = await subject(
      `
  /** @type {import('eslint').Linter.Config} should be single line */
  const config = {
    // ...
  };
`,
      {
        jsdocCommentLineStrategy: "keep",
      },
    );
    expect(result).toMatchSnapshot();
  });

  test("keep multi", async () => {
    const result1 = await subject(
      `
  /**
   * @type {import('eslint').Linter.Config} should be multiline
   */
  const config = {
    // ...
  };
`,
      {
        jsdocCommentLineStrategy: "keep",
      },
    );

    expect(result1).toMatchSnapshot();
  });
  test("singleLine ", async () => {
    const result2 = await subject(
      `
  /**
   * @type {import('eslint').Linter.Config} should be single
   */
  const config = {
    // ...
  };
`,
      {
        jsdocCommentLineStrategy: "singleLine",
      },
    );
    expect(result2).toMatchSnapshot();
  });
  test("multiline ", async () => {
    const result3 = await subject(
      `
  /** @type {import('eslint').Linter.Config} should be multiline */
  const config = {
    // ...
  };
`,
      {
        jsdocCommentLineStrategy: "multiline",
      },
    );
    expect(result3).toMatchSnapshot();
  });
});
