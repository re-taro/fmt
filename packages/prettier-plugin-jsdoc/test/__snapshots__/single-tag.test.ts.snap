// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Comment Line Strategy > keep multi 1`] = `
"/**
 * @type {import(\\"eslint\\").Linter.Config} should Be multiline
 */
const config = {
  // ...
};
"
`;

exports[`Comment Line Strategy > keep single 1`] = `
"/** @type {import(\\"eslint\\").Linter.Config} should Be single line */
const config = {
  // ...
};
"
`;

exports[`Comment Line Strategy > multiline  1`] = `
"/**
 * @type {import(\\"eslint\\").Linter.Config} should Be multiline
 */
const config = {
  // ...
};
"
`;

exports[`Comment Line Strategy > singleLine  1`] = `
"/** @type {import(\\"eslint\\").Linter.Config} should Be single */
const config = {
  // ...
};
"
`;

exports[`single tag 1`] = `
"/** @param {string} param0 Description */
function fun(param0) {}

export const SubDomain = {
  /** @returns {import(\\"axios\\").AxiosResponse<import(\\"../types\\").SubDomain>} */
  async subDomain(subDomainAddress) {},
};
"
`;
